tools/autograd/templates/python_torch_functions.cpp

```diff
@@ -465,6 +467,109 @@ static PyObject * THPVariable_get_device(PyObject* self_, PyObject* args, PyObje
   END_HANDLE_TH_ERRORS
 }
 
+static PyObject * THPVariable_from_buffer(PyObject* self_, PyObject* args, PyObject* kwargs)
+{
+  HANDLE_TH_ERRORS
+  static PythonArgParser parser({
+    "from_buffer(PyObject* buffer, ScalarType dtype=None, int64_t count=-1, int64_t offset=0, *, Device? device=None, bool requires_grad=False)",
+  }, /*traceable=*/false);
+
+  ParsedArgs<6> parsed_args;
+  auto r = parser.parse(args, kwargs, parsed_args);
+
+  if (r.idx == 0) {
+      auto buffer = r.pyobject(0);
+      auto dtype = r.scalartype(1);
+      auto count = r.toInt64(2);
+      auto offset = r.toInt64(3);
+      auto device_opt = r.deviceOptional(4);
+      auto requires_grad = r.toBool(5);
+
+      auto elsize = at::elementSize(dtype);
+      size_t actual_count = 0;
+
+      Py_buffer view;
+
+      if (PyObject_GetBuffer(buffer, &view, PyBUF_SIMPLE) < 0) {
+          return nullptr;
+      }
+
+      auto len = view.len;
+      auto buf = view.buf;
+      auto obj = view.obj;
+      PyBuffer_Release(&view);
+
+      if (len == 0 || count == 0) {
+        PyErr_SetString(
```

Comment:
The problem is that: given that I `Py_INCREF(obj)` before this check, I think I should `Py_DECREF(obj)` in case this check fails. So, I guess the ifs are needed here.

Reply:
If a variable has an owning reference to a PyObject, then you want to keep it in a `THPObjectPtr`. This will do the decref automatically when the object is destroyed, avoiding such leak.

You can see a few use of this in `torch/csrc/Module.cpp` in particular.

Reply:
Hmm. I refactored the code a bit, tracking whether an error has occurred or not with an `c10::optional<std::string>` (which allows me to correctly call `Py_DECREF` accordingly). Do you think it's better to undo it, and just use `TORCH_CHECK_VALUE`?

Reply:
Let me check the code.